cocos/gfx/webgpu/webgpu-command-buffer.ts

```diff
@@ -0,0 +1,740 @@
+                const cmd = this._webGPUAllocator!.updateBufferCmdPool.alloc(WebGPUCmdUpdateBuffer);
+                let buffSize = 0;
+                let buff: BufferSource | null = null;
+
+                // TODO: Have to copy to staging buffer first to make this work for the execution is deferred.
+                // But since we are using specialized primary command buffers in WebGL backends, we leave it as is for now
+                if (buffer.usage & BufferUsageBit.INDIRECT) {
+                    buff = data;
+                } else {
+                    if (size !== undefined) {
+                        buffSize = size;
+                    } else {
+                        buffSize = (data as ArrayBuffer).byteLength;
+                    }
+                    buff = data;
+                }
+                const device = WebGPUDeviceManager.instance;
+                const nativeDevice = (device as WebGPUDevice).nativeDevice;
+                nativeDevice?.queue.writeBuffer(gpuBuffer.glBuffer!, gpuBuffer.glOffset, buff as ArrayBuffer);
+
+                // cmd.gpuBuffer = gpuBuffer;
+                // cmd.buffer = buff;
+                // cmd.offset = (offset !== undefined ? offset : 0);
+                // cmd.size = buffSize;
+                // this.cmdPackage.updateBufferCmds.push(cmd);
+                //
+                // this.cmdPackage.cmds.push(WebGPUCmd.UPDATE_BUFFER);
+            }
+        } else {
+            console.error('Command \'updateBuffer\' must be recorded outside a render pass.');
+        }
+    }
+
+    public copyBuffersToTexture(buffers: ArrayBufferView[], texture: Texture, regions: BufferTextureCopy[]) {
+        if (this._type === CommandBufferType.PRIMARY && !this._isInRenderPass
+            || this._type === CommandBufferType.SECONDARY) {
+            const gpuTexture = (texture as WebGPUTexture).gpuTexture;
+            if (gpuTexture) {
+                const cmd = this._webGPUAllocator!.copyBufferToTextureCmdPool.alloc(WebGPUCmdCopyBufferToTexture);
+                cmd.gpuTexture = gpuTexture;
+                cmd.regions = regions;
+                // TODO: Have to copy to staging buffer first to make this work for the execution is deferred.
+                // But since we are using specialized primary command buffers in WebGL backends, we leave it as is for now
+                cmd.buffers = buffers;
+
+                this.cmdPackage.copyBufferToTextureCmds.push(cmd);
+                this.cmdPackage.cmds.push(WebGPUCmd.COPY_BUFFER_TO_TEXTURE);
+            }
+        } else {
+            console.error('Command \'copyBufferToTexture\' must be recorded outside a render pass.');
+        }
+    }
+
+    public execute(cmdBuffs: CommandBuffer[], count: number) {
+        for (let i = 0; i < count; ++i) {
+            const WebGPUCmdBuff = cmdBuffs[i] as WebGPUCommandBuffer;
+
+            for (let c = 0; c < WebGPUCmdBuff.cmdPackage.beginRenderPassCmds.length; ++c) {
```

Comment:
It is better to compute length only once, such as 
```
for (let c = 0, len = WebGPUCmdBuff.cmdPackage.beginRenderPassCmds.length; c < len; ++c)
```

Please check all usages.

Reply:
Modified